### PR TITLE
Improve predictability of DataQuery, DataID, and dependency tree

### PR DESCRIPTION
**Pull request by djhoese**
Disclaimer: This PR's implementation is *very* rough at the time of writing.

This PR includes major changes to key parts of Satpy in order to resolve some inconsistencies noticed by users over the years. The high-level concepts that have been changed or updated are:

1. DataQuery objects are now only equal to DataID objects that match *all* of the queries keys. Previously only the shared keys were compared. This old way meant that a DataQuery could match a DataID that didn't contain all the necessary information requested by the query.
2. The "resolution" DataID key was not not transitive for all ID key sets. So for "default" ID keys it was False and for coordinate and minimal sets it was transitive. It made the most sense to set it to False. That is, a modified dataset is not *required* to have all dependencies be of the same resolution.
3. Add and refactor a lot of tests regarding DataQuery and DataID comparisons.
4. It should be possible to load a composite with two different sets of inputs (ex. `DataQuery(name="comp", resolution=500), DataQuery(name="comp", resolution=1000)`).

### Remaining work

* There are a lot of edge cases that need to be worked out. The biggest one is what happens when a DataQuery has a list of possible options. That is not handled in a lot of my dependency tree stuff.
* See "Hindsight" below.
* Refactoring
* More explicit tests

### Hindsight

For high-level change 1 above, I'm starting to think this was the wrong change or at least that the previous behavior had a good point. That is, if a user creates a query with a lot of keys to apply to many DataIDs from different sources, then not all DataIDs should be required to have all those keys. For example, if I specify a polarization in my query, then I don't think all composites or rather composite dependencies would be able to match that. There are currently no tests to verify this.

 - [ ] Closes #xxxx 
 - [x] Tests added 
 - [ ] Fully documented 
 - [ ] Add your name to `AUTHORS.md` if not there already

### CODE REVIEW

File: satpy/dataset/dataid.py [hunk 1]
```diff
@@ -732,21 +481,85 @@ def create_filtered_query(dataset_key, filter_query):
...
+def update_id_with_query(orig_id: DataID, query: DataQuery) -> DataID:
+    """Update a DataID with additional info from a query used to find it."""
+    query_dict = query.to_dict()
+    if not query_dict:
+        return orig_id
+
+    new_id_dict = orig_id.to_dict()
+    orig_id_keys = orig_id.id_keys
+    for query_key, query_val in query_dict.items():
+        # XXX: What if the query_val is a list?
```
Comment (djhoese) [inline satpy/dataset/dataid.py:512]: This is my main task remaining. I'm really not sure how I want to treat this. If you ask for a composite as `DataQuery(name="some_comp", resolution=[500, 1000])`, what do you set in the `DataID`? This is before we know what dependencies were found. So does the composite `DataID` become `DataID(name="some_comp")` with no resolution, resolution 500, or resolution 1000?

Comment (mraspaud) [reply satpy/dataset/dataid.py:512]: imo it should get the final resolution. So if the generated composite has a 500m resolution, the dataid should carry that.

File: satpy/dataset/dataid.py [hunk 2]
```diff
@@ -732,21 +481,85 @@ def create_filtered_query(dataset_key, filter_query):
...
+        if new_id_dict.get(query_key) is None:
+            new_id_dict[query_key] = query_val
+    # don't replace ID key information if we don't have to
+    id_keys = orig_id_keys if all(key in orig_id_keys for key in new_id_dict) else default_id_keys_config
+    new_id = DataID(id_keys, **new_id_dict)
+    return new_id
+
+
+def _keys_to_compare(sdict: dict, odict: dict, o_is_id: bool, shared_keys: bool) -> set:
```
Comment (mraspaud) [inline satpy/dataset/dataid.py:521]: what does `o_is_id` stand for?

File: satpy/dataset/dataid.py [hunk 3]
```diff
@@ -732,21 +481,85 @@ def create_filtered_query(dataset_key, filter_query):
...
+        if o_is_id:
+            return oval in sval
+
+        # we're matching against a DataQuery who could have its own list
+        if not isinstance(oval, list):
+            oval = [oval]
+        s_in_o = any(_sval in oval for _sval in sval)
+        o_in_s = any(_oval in sval for _oval in oval)
+        return s_in_o or o_in_s
+    return oval == sval
```
Comment (mraspaud) [inline satpy/dataset/dataid.py:565]: It’s probably my doing, but could we take the opportunity to find better/longer names for sval, oval, sdict, odict, etc?

File: satpy/dependency_tree.py [hunk 4]
```diff
@@ -424,6 +430,7 @@ def _find_compositor(self, dataset_key, query):
         # one or more modifications if it has modifiers see if we can find
         # the unmodified version first
 
+        orig_query = dataset_key
```
Comment (mraspaud) [inline satpy/dependency_tree.py:433]: 1. are we sure this is a query, not an id?
2. should we rename the function argument instead?

File: satpy/dependency_tree.py [hunk 5]
```diff
@@ -488,14 +496,43 @@ def _promote_query_to_modified_dataid(self, query, dep_key):
...
                 continue
 
+        if key.get("name", default="*") != "*" and len(key.to_dict()) == 1:
+            # the query key is just the name and still couldn't be found
+            raise KeyError("Could not find compositor '{}'".format(key))
+
+        # Get the generic version of the compositor (by name only)
+        # then save our new version under the new name
+
+        return self._get_compositor_by_name(key)
```
Comment (mraspaud) [inline satpy/dependency_tree.py:514]: What use cases in this covering?

File: satpy/tests/test_dataset.py [hunk 6]
```diff
@@ -643,26 +618,129 @@ def test_id_filtering(self):
...
+        ("dq1", "dq2"),
+        [
+            (DataQuery(name="1"), DataQuery(name="1")),
+            (DataQuery(name="1", resolution="*"), DataQuery(name="1")),
+            (DataQuery(name="1", resolution="*"), DataQuery(name="1", resolution=500)),
+            (DataQuery(name="1", resolution=500), DataQuery(name="1", resolution="*")),  # opposite order
+            (DataQuery(), DataQuery()),
+            (DataQuery(name="1", resolution=[250, 500]), DataQuery(name="1", resolution=[500, 750])),  # opposite order
+            (DataQuery(name="1", resolution=500), DataQuery(name="1", resolution=[500, 750])),  # opposite order
+            (DataQuery(name="1", resolution=[250, 500]), DataQuery(name="1", resolution=500)),  # opposite order
```
Comment (mraspaud) [inline satpy/tests/test_dataset.py:673]: These look surprising?

**Review by mraspaud — COMMENTED** [review]
Ok, I made a first pass. Overall, thanks a lot for the clarifications and refactorings, the code reads better now.
I have comments inline, or rather mostly questions because I’m not following everything :)

File: satpy/dataset/dataid.py [hunk 7]
```diff
@@ -732,21 +481,85 @@ def create_filtered_query(dataset_key, filter_query):
...
     if filter_query is not None:
         for key, value in filter_query.items():
             if value != "*":
                 ds_dict.setdefault(key, value)
 
 
-def _create_id_dict_from_any_key(dataset_key):
-    try:
+def _create_id_dict_from_any_key(dataset_key: DataQuery | DataID | str | numbers.Number) -> dict[str, Any]:
+    if hasattr(dataset_key, "to_dict"):
```
[the remark this answers is not in the document]
Comment (djhoese) [reply satpy/dataset/dataid.py:492]: If I remember correctly I changed this because mypy was getting mad about the typing and the `try/except` being hard to parse. It might have also been CodeScene. I knew you would prefer try/except here, but the linter's didn't like it so I thought it was OK. I can do some type ignoring if you'd still prefer the try/except.

File: satpy/dataset/dataid.py (hunk 2 above)
Comment (djhoese) [reply satpy/dataset/dataid.py:521]: other is DataID or at least looks like one. I could give the variable a longer name for sure. I went back and forth on how to handle the different cases, but at the end of the day my sanity was preserved by doing simple if/else statements on the DataID-ness of "other". Now that most of the logic is stabilizing I could revisit this. Ideas welcome.

File: satpy/dependency_tree.py (hunk 4 above)
Comment (djhoese) [reply satpy/dependency_tree.py:433]: I'm honestly not sure, but I think so. This is reminding me that I wanted to add more type annotations to this stuff for this reason, but forgot after needing some last minute changes in other parts of the code. If I remember correctly the information passed by the user gets turned into a DataQuery in one of the top-level tree methods and gets passed around from there.

File: satpy/tests/test_dataset.py (hunk 6 above)
Comment (djhoese) [reply satpy/tests/test_dataset.py:673]: This is a case of the code written first then the tests second. More importantly, this is testing code that I didn't write, but I could tell this functionality was intended. That is, I believe we support users specifying a list of things in their DataQuery (and in the filter kwargs passed to `Scene.load`) and the results should/could match any of the things in the list.

Or is there something else surprising about these. I'm mostly making sure that regardless of which DataQuery object's `__eq__` is called (the first or the second) that things are still matched. I'm not sure if the `# opposite order` comment at the end of these 3 lines is valid. Most likely it is an artifact of copy/pasting the earlier test. Or maybe the comment doesn't apply to the 3rd to last case, but does for the last 2 cases.

File: satpy/dataset/dataid.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mraspaud) [reply satpy/dataset/dataid.py:492]: that fine…

File: satpy/dependency_tree.py (hunk 5 above)
Comment (djhoese) [reply satpy/dependency_tree.py:514]: Ah the name of this method is incorrect after more refactoring was done. So the code above this is when the compositor exactly matches what the user asked for. This is usually the common case of just a compositor by name but could also include other filtering parameters like `resolution` or `calibration` or whatever else was asked for.

The method `_get_compositor_by_name` first gets the base compositor by just the name and then if more than one compositor matches it filters down using the other properties of the provided `DataQuery`. The cases where a compositor name returns more than one compositor would be if there are varying resolutions or calibrations for a single compositor. So again, not a common use case, but it was in the tests already.

File: satpy/dataset/dataid.py [hunk 8]
```diff
@@ -512,35 +263,64 @@ def __getitem__(self, key):
...
+        are shared between them and are equal. A DataQuery is considered
+        equal to a DataID if all elements in the query are equal to those
+        elements in the DataID. The DataID is still considered equal if it
+        contains additional elements. Any DataQuery elements with the value
+        ``"*"`` are ignored.
+
+        """
+        return self.equal(other, shared_keys=False)
+
+    def equal(self, other: Any, shared_keys: bool = False) -> bool:
```
[the remark this answers is not in the document]
Comment (djhoese) [reply satpy/dataset/dataid.py:279]: Not really as the ability to get shared keys (keys to match) only happens after we've extracted the dict version of each object. Additionally, this is a top-level public method now (I don't remember if it needs to be) so doing low-level stuff with the keys seems like too much work for a high-level user-facing method and would have to be repeated anywhere `shared_keys=True` is used. I could split it I suppose into two methods that call one shared method, but then again it is a lot of extra code for a single use.

File: satpy/dataset/dataid.py (hunk 3 above)
Comment (djhoese) [reply satpy/dataset/dataid.py:565]: I really thought I answered a bunch of these comments already, but apparently not...

Yes, I can try to find better names. I think `s` is self or maybe source? And `o` is other. I'm not entirely upset with them as they are now, but yeah not super clear.

File: satpy/dataset/dataid.py (hunk 1 above)
Comment (djhoese) [reply satpy/dataset/dataid.py:512]: And that may happen eventually, but we don't know that when this function is called. It is called here:

https://github.com/pytroll/satpy/blob/c45ed8d831284405826d53bc919b5f09ccfa2d00/satpy/dependency_tree.py#L448

There are the two separate processes of getting the compositor (the class instance that will be called) and the composite DataArray (generated by calling the compositor). This function is called to give the compositor an identity so it can be referred to for generating, but also if it is needed by some other compositor configuration (YAML prerequisites) or the user requested it. So overall the DataID generated by this function is "temporary" in that once composites are generated this DataID may be overwritten by the `__call__` of the compositor.

Additionally, at least with the current calling order of things, we don't even know which prerequisites are going to be used when this function is called. And it seems the CompositorNode where this is DataID is used is needed before prerequisites and optionals are determined because we add the Nodes of those prereqs/optionals to the CompositorNode's children. There might be a way to get or hack around that, but I'm not sure.

So back to the original question:

> If you ask for a composite as DataQuery(name="some_comp", resolution=[500, 1000]), what do you set in the DataID?

We don't know the prereqs or optionals and we're just updating the original request of "some_comp" with *some* resolution, but what one?

**Comment by djhoese** [comment]
Ok I've been spending some time trying to get a better understanding of composite and modifier ID handling. Basically I wanted to be more confident in my fixes, add tests for edge cases, and resolve my TODO I mentioned in [this comment](https://github.com/pytroll/satpy/pull/3018#discussion_r1892683862). It comes down to compositors (and the Node objects wrapping them) and composites (the loaded DataArray) being separate identifiers, but treated as the same thing.

## Series of events

1. User calls `Scene.load` and [the Scene determines](https://github.com/pytroll/satpy/blob/3990f37a36a53bcdaa1d1f144c42728db4ca8dbd/satpy/scene.py#L1450) what from the request is still needed. I'm still not sure how this code works between strings/DataIDs/DataQuerys and the DataIDs in the `self._datasets` but I'll figure that out.
2. The Scene updates the dependency tree so it knows where and how to create the datasets requested by the user.
3. The dependency tree, in the case of a compositor, will find the compositor usually by name, and let the compositor say what DataID it has by accessing the `.id` property of the Compositor object (`compositor.id` - see usage [here](https://github.com/djhoese/satpy/blob/c45ed8d831284405826d53bc919b5f09ccfa2d00/satpy/node.py#L163-L165)). The dependency tree now stores the compositor by that DataID.
4. The Scene gets the compositor nodes to be loaded (trunk nodes):

   https://github.com/pytroll/satpy/blob/3990f37a36a53bcdaa1d1f144c42728db4ca8dbd/satpy/scene.py#L1541-L1542

5. The Scene checks if the dataset in the DependencyTree already exists in the Scene's container of created datasets:

   https://github.com/pytroll/satpy/blob/3990f37a36a53bcdaa1d1f144c42728db4ca8dbd/satpy/scene.py#L1533-L1537

6. Any compositor nodes from the dependency tree still needing loading is generated, but we first check if it is contained in the Scene's container...again:

   https://github.com/pytroll/satpy/blob/3990f37a36a53bcdaa1d1f144c42728db4ca8dbd/satpy/scene.py#L1564-L1566

   Same check as step 5 above, but more strict as `.contains` must match the DataID exactly while step 5 is a "best match".

7. After the DataArray is generated from the compositor we create a DataID from it's `.attrs` and update the Scene's wishlist and the dependency trees nodes with this new ID:

   https://github.com/pytroll/satpy/blob/3990f37a36a53bcdaa1d1f144c42728db4ca8dbd/satpy/scene.py#L1611-L1618

## Problems and Difficulties

1. In step 3 above, the compositor's DataID and the resulting DataArray might not be the same. The compositor object's `.id` is not able to know what version of it's prerequisites will be used (the dependency tree hasn't determined that yet) and therefore can't use it to influence the DataID of the compositor. Historically we've allowed this so the compositor developer can do at-runtime updates to the DataID parameters to make them more accurate (ex. update "resolution") in the compositor's `__call__`.
2. Due to problem 1, the checks in step 1, 5, and 6 may not be accurate depending on how the compositor updates the DataID for the resulting DataArray.
3. In `main`, the compositor's ID does not take into account the extra query parameters passed by the user (either from the DataQuery from the user, or the `**kwargs` passed like `resolution=[500, 1000]`). This means a user could request the same composite multiple ways that should result in different inputs being used, but the DependencyTree would use the same compositor for all of them. In this PR I try to solve that, but there are problems [with that as well](https://github.com/pytroll/satpy/pull/3018#discussion_r1892683862).
4. The whole compositor ID re-assigning is very unsettling.

## Solutions?

:man_shrugging: 

Overall (I think in `main` and this PR) if a compositor's `__call__` modifies the resulting `DataID` in a way that conflicts with the previous compositor ID (changes a value that already existed) or makes it not match the users original request, then the resulting behavior is not well defined. This might just be something we have to deal with.

**Comment by djhoese** [comment]
I brought up some of these issues on slack and proposed a new method for compositor classes that would take the dependency DataIDs as input and produce the resulting DataID. After sending that message and looking at the code I'm realizing that the existing `.id` property of compositors could be reused/redefined and that if compositor instantiation (calling of `__init__`) was delayed until the dependency tree needed it, then the dependency tree could pass prerequisites and optional prerequisites as DataIDs instead of DataQuery objects like it does now. The `.id` property could then use these DataIDs to determine a DataID for the compositor and composite DataArray that will be generate with the `__call__` later on.

The one remaining question is what are the rules for merging of the input DataIDs and what happens when there are conflicts. Like if 2 of 3 of the inputs have resolution=500 then the resulting DataID would have `resolution=500`, but what if all 3 of the input DataIDs had resolution=500, then the resulting compositor DataID would also be `resolution=500`. Two different sets of inputs resulting in the same identifier.
